onnxruntime/core/providers/nnapi/nnapi_builtin/builders/model_builder.cc

```diff
@@ -24,10 +26,16 @@ namespace onnxruntime {
 namespace nnapi {
 
 ModelBuilder::ModelBuilder(const GraphViewer& graph_viewer)
-    : nnapi_(NnApiImplementation()), graph_viewer_(graph_viewer), shaper_{graph_viewer} {}
-
-int32_t ModelBuilder::GetNNAPIFeatureLevel() const {
-  return nnapi_ ? static_cast<int32_t>(nnapi_->nnapi_runtime_feature_level) : 0;
+    : nnapi_(NnApiImplementation()), graph_viewer_(graph_viewer), shaper_{graph_viewer} {
+  if (auto st = GetTargetDevices(nnapi_, target_device_option_, nnapi_target_devices_, nnapi_target_devices_detail_);
+      !st.IsOK()) {
```

Comment:
Can we separate out all the logic around calculating the effective level and providing diagnostic output into a helper class? It does not seem like a core responsibility of a class that builds a model.

Instead the ModelBuilder should use the helper to populate a property with the effective level in the ctor, make that value accessible to its users, and call a (probably static) method of the helper to do things like the diagnostic dump of nodes using nnapi-reference.

Reply:
Level calculation has been placed on EP level. So, it's more consistent to selecting Ops and running/building ops.
 
diagnostic message can be dumped by current devices. Multiple calling to NNAPI-API is not happened ever.